File: src/splitter.py
##!/opt/local/bin/python3.4
##!/usr/bin/python -O

# -*- coding: iso-8859-15 -*-

# This code is distributed under the GNU General Public License (see
# THE_GENERAL_GNU_PUBLIC_LICENSE.txt for extending this information).
# Copyright (C) 2014, the P2PSP team.
# http://www.p2psp.org

# PYTHON_ARGCOMPLETE_OK

# {{{ Imports

from __future__ import print_function
import time
import sys
import socket
import threading
import struct

from core.common import Common
from core._print_ import _print_
from core.color import Color
from core.splitter_ims import Splitter_IMS
from core.splitter_dbs import Splitter_DBS
from core.splitter_acs import Splitter_ACS
from core.splitter_nts import Splitter_NTS

try:
    import colorama                       # Enable console color using ANSI codes in Windows
except ImportError:
    pass

import argparse
try:
    import argcomplete                    # Bash tab completion for argparse in Unixes
except ImportError:
    pass

# }}}

class Splitter():

    def __init__(self):

        # {{{ colorama.init()

        try:
            colorama.init()
        except Exception:
            pass

        # }}}

        # {{{ Running in debug/release mode

        _print_("Running in", end=' ')
        if __debug__:
            print("debug mode")
        else:
            print("release mode")

        # }}}

        # {{{ Arguments handling

        parser = argparse.ArgumentParser(description='This is the splitter node of a P2PSP team.  The splitter is in charge of defining the Set or Rules (SoR) that will control the team. By default, DBS (unicast transmissions) will be used.')
        #parser.add_argument('--splitter_addr', help='IP address to serve (TCP) the peers. (Default = "{}")'.format(Splitter_IMS.SPLITTER_ADDR)) <- no ahora
        parser.add_argument('--buffer_size', help='size of the video buffer in blocks. Default = {}.'.format(Splitter_IMS.BUFFER_SIZE))
        parser.add_argument('--channel', help='Name of the channel served by the streaming source. Default = "{}".'.format(Splitter_IMS.CHANNEL))
        parser.add_argument('--chunk_size', help='Chunk size in bytes. Default = {}.'.format(Splitter_IMS.CHUNK_SIZE))
        parser.add_argument('--header_size', help='Size of the header of the stream in chunks. Default = {}.'.format(Splitter_IMS.HEADER_SIZE))
        parser.add_argument('--max_chunk_loss', help='Maximum number of lost chunks for an unsupportive peer. Makes sense only in unicast mode. Default = {}.'.format(Splitter_DBS.MAX_CHUNK_LOSS))
        parser.add_argument('--max_number_of_monitor_peers', help='Maxium number of monitors in the team. The first connecting peers will automatically become monitors. Default = "{}".'.format(Splitter_DBS.MONITOR_NUMBER))
        parser.add_argument('--mcast_addr', help='IP multicast address used to serve the chunks. Makes sense only in multicast mode. Default = "{}".'.format(Splitter_IMS.MCAST_ADDR))
        parser.add_argument('--port', help='Port to serve the peers. Default = "{}".'.format(Splitter_IMS.PORT))
        parser.add_argument('--source_addr', help='IP address or hostname of the streaming server. Default = "{}".'.format(Splitter_IMS.SOURCE_ADDR))
        parser.add_argument('--source_port', help='Port where the streaming server is listening. Default = {}.'.format(Splitter_IMS.SOURCE_PORT))
        parser.add_argument("--IMS", action="store_true", help="Uses the IP multicast infrastructure, if available. IMS mode is incompatible with ACS, LRS, DIS and NTS modes.")
        parser.add_argument("--NTS", action="store_true", help="Enables NAT traversal.")
        parser.add_argument("--ACS", action="store_true", help="Enables Adaptive Chunk-rate.")
        parser.add_argument("--LRS", action="store_true", help="Enables Lost chunk Recovery.")
        parser.add_argument("--DIS", action="store_true", help="Enables Data Integrity check.")
        parser.add_argument('--strpe', nargs='+', type=str, help='Selects STrPe model for DIS')
        parser.add_argument('--strpeds', nargs='+', type=str, help='Selects STrPe-DS model for DIS')
        parser.add_argument('--strpeds_majority_decision', help='Sets majority decision ratio for STrPe-DS model.')
        parser.add_argument('--strpe_log', help='Logging STrPe & STrPe-DS specific data to file.')
        parser.add_argument('--TTL', help='Time To Live of the multicast messages. Default = {}.'.format(Splitter_IMS.TTL))
        
        try:
            argcomplete.autocomplete(parser)
        except Exception:
            pass
        args = parser.parse_args()
        #args = parser.parse_known_args()[0]

        _print_("My IP address is =", socket.gethostbyname(socket.gethostname()))
        
        if args.buffer_size:
            Splitter_IMS.BUFFER_SIZE = int(args.buffer_size)
        _print_("Buffer size =", Splitter_IMS.BUFFER_SIZE)

        if args.channel:
            Splitter_IMS.CHANNEL = args.channel
        _print_("Channel = \"" + Splitter_IMS.CHANNEL + "\"")

        if args.chunk_size:
            Splitter_IMS.CHUNK_SIZE = int(args.chunk_size)
        _print_("Chunk size =", Splitter_IMS.CHUNK_SIZE)

        if args.header_size:
            Splitter_IMS.HEADER_SIZE = int(args.header_size)
        _print_("Header size =", Splitter_IMS.HEADER_SIZE)

        if args.port:
            Splitter_IMS.PORT = int(args.port)
        _print_("Listening port =", Splitter_IMS.PORT)

        if args.source_addr:
            Splitter_IMS.SOURCE_ADDR = socket.gethostbyname(args.source_addr)
        _print_("Source address = ", Splitter_IMS.SOURCE_ADDR)

        if args.source_port:
            Splitter_IMS.SOURCE_PORT = int(args.source_port)
        _print_("Source port =", Splitter_IMS.SOURCE_PORT)

        if args.IMS:
            _print_("IP multicast (IMS) mode selected")

            if args.mcast_addr:
                    Splitter_IMS.MCAST_ADDR = args.mcast_addr
            _print_("Multicast address =", Splitter_IMS.MCAST_ADDR)

            if args.TTL:
                    Splitter_IMS.TTL = args.TTL
            _print_("Multicast TTL =", Splitter_IMS.TTL)

            splitter = Splitter_IMS()
            splitter.peer_list = [] # No peer_list is used in IMS.
                
        else:
            _print_("IP unicast mode selected")

            if args.max_chunk_loss:
                Splitter_DBS.MAX_CHUNK_LOSS = int(args.max_chunk_loss)
            _print_("Maximun chunk loss =", Splitter_DBS.MAX_CHUNK_LOSS)

            if args.max_number_of_monitor_peers:
                Splitter_DBS.MONITOR_NUMBER = int(args.max_number_of_monitor_peers)
            _print_("Maximun number of monitor peers =", Splitter_DBS.MONITOR_NUMBER)

            splitter = Splitter_DBS()

            if args.NTS:
                splitter = Splitter_NTS(splitter)
                _print_("NTS enabled")

            if args.ACS:
                splitter = Splitter_ACS(splitter)
                _print_("ACS enabled")

            if args.LRS:
                from core.splitter_lrs import Splitter_LRS
                splitter = Splitter_LRS(splitter)
                _print_("LRS enabled")

            if args.DIS:
                from splitter_strpe import StrpeSplitter
                from splitter_strpeds import StrpeDsSplitter
                _print_("DIS enabled")
                if args.strpe:
                    splitter = Splitter_strpe(splitter)
                    print("strpe mode selected")
                    for peer in args.strpe:
                        splitter.add_trusted_peer(peer)
                if args.strpeds:
                    splitter = StrpeSplitter(splitter)
                    _print_("strpeds mode selected")
                    for peer in args.strpeds:
                        splitter.add_trusted_peer(peer)
                    if args.strpeds_majority_decision:
                        _print_("strpeds_majority_decision mode selected")
                        splitter = Splitter_strpeds_majority_decision(splitter)
                        splitter.setMajorityRatio(float(args.strpeds_majority_decision))
                if args.strpe_log:
                    splitter.LOGGING = True
                    splitter.LOG_FILE = open(strpe_log, 'w', 0)

            #splitter = Splitter_ACS()
#            if (args.strpe):
#                splitter = self.init_strpe_splitter('strpe', args.strpe, args.strpe_log)
#            elif (args.strpeds):
#                splitter = self.init_strpe_splitter('strpeds', args.strpeds, args.strpe_log)
#                if args.strpeds_majority_decision:
#                    splitter.setMajorityRatio(float(args.strpeds_majority_decision))
#            else:
#                splitter = Splitter_LRS()

        # }}}

        # {{{ Run!

        splitter.start()

        # {{{ Prints information until keyboard interruption

        print("         | Received  | Sent      | Number       losses/ losses")
        print("    Time | (kbps)    | (kbps)    | peers (peer) sents   threshold period kbps")
        print("---------+-----------+-----------+-----------------------------------...")

        last_sendto_counter = splitter.sendto_counter
        last_recvfrom_counter = splitter.recvfrom_counter

        while splitter.alive:
            try:
                time.sleep(1)
                chunks_sendto = splitter.sendto_counter - last_sendto_counter
                kbps_sendto = (chunks_sendto * splitter.CHUNK_SIZE * 8) / 1000
                chunks_recvfrom = splitter.recvfrom_counter - last_recvfrom_counter
                kbps_recvfrom = ( chunks_recvfrom * splitter.CHUNK_SIZE * 8) / 1000
                last_sendto_counter = splitter.sendto_counter
                last_recvfrom_counter = splitter.recvfrom_counter
                sys.stdout.write(Color.none)
                _print_("|" + repr(int(kbps_recvfrom)).rjust(10) + " |" + repr(int(kbps_sendto)).rjust(10), end=" | ")
                #print('%5d' % splitter.chunk_number, end=' ')
                sys.stdout.write(Color.cyan)
                print(len(splitter.peer_list), end=' ')
                if not __debug__:
                    counter = 0
                for p in splitter.peer_list:
                    if not __debug__:
                        if counter > 10:
                            break
                        counter += 1
                    sys.stdout.write(Color.blue)
                    print(p, end= ' ')
                    sys.stdout.write(Color.red)
                    print(str('%3d' % splitter.losses[p]) + '/' + str('%3d' % chunks_sendto), splitter.MAX_CHUNK_LOSS, end=' ')
                    if splitter is Splitter_ACS:
                        try:
                            sys.stdout.write(Color.yellow)
                            print('%3d' % splitter.period[p], end= ' ')
                            sys.stdout.write(Color.purple)
                            print(repr((splitter.number_of_sent_chunks_per_peer[p] * splitter.CHUNK_SIZE * 8) / 1000).rjust(10), end = ' ')
                            splitter.number_of_sent_chunks_per_peer[p] = 0
                        except KeyError as e:
                            print("!", e, "--")
                            print(splitter.period[p])
                            pass
                    sys.stdout.write(Color.none)
                    print('', end=' ')
                print()

            except KeyboardInterrupt:
                print('Keyboard interrupt detected ... Exiting!')

                # Say to daemon threads that the work has been finished,
                splitter.alive = False

                # Wake up the "moderate_the_team" daemon, which is
                # waiting in a recvfrom().
                if not args.IMS:
                    #splitter.say_goodbye(("127.0.0.1", splitter.PORT), splitter.team_socket)
                    splitter.team_socket.sendto(b'', ("127.0.0.1", splitter.PORT))

                # Wake up the "handle_arrivals" daemon, which is waiting
                # in an accept().
                sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                sock.connect(("127.0.0.1", splitter.PORT))
                sock.recv(struct.calcsize("4sH")) # Multicast channel
                sock.recv(struct.calcsize("H")) # Header size
                sock.recv(struct.calcsize("H")) # Chunk size
                sock.recv(splitter.CHUNK_SIZE*splitter.HEADER_SIZE) # Header
                sock.recv(struct.calcsize("H")) # Buffer size
                sock.recv(struct.calcsize("4sH")) # Endpoint
                sock.recv(struct.calcsize("B")) # Magic flags
                if args.IMS:
                    number_of_peers = 0
                else:
                    number_of_peers = socket.ntohs(struct.unpack("H", sock.recv(struct.calcsize("H")))[0])
                    print("Number of peers =", number_of_peers)
                # Receive the list
                while number_of_peers > 0:
                    sock.recv(struct.calcsize("4sH"))
                    number_of_peers -= 1

                # Breaks this thread and returns to the parent process
                # (usually, the shell).
                break

            # }}}

        # }}}

    # {{{# -COM-

    def init_strpe_splitter(self, type, trusted_peers, log_file = None):
        if type == 'strpe':
            re = StrpeSplitter()
        if type == 'strpeds':
            re = StrpeDsSplitter()
        for peer in trusted_peers:
            re.add_trusted_peer(peer)
        if log_file != None:
            re.LOGGING = True
            re.LOG_FILE = open(log_file, 'w', 0)
        return re

    # }}}


x = Splitter()

File: src/splitter.cc
//
//  splitter.cpp
//  P2PSP
//
//  This code is distributed under the GNU General Public License (see
//  THE_GENERAL_GNU_PUBLIC_LICENSE.txt for extending this information).
//  Copyright (C) 2016, the P2PSP team.
//  http://www.p2psp.org
//

#include <iostream>
#include <memory>
#include "core/common.h"
#include "core/splitter_ims.h"
#include "core/splitter_dbs.h"
#include "core/splitter_acs.h"
#include "core/splitter_lrs.h"
#include "core/splitter_strpe.h"
#include "core/splitter_nts.h"
#include <boost/program_options/parsers.hpp>
#include <boost/program_options/variables_map.hpp>
#include <signal.h>
#include "util/trace.h"

// TODO: LOG fails if splitter is defined outside the main
// p2psp::SplitterSTRPE splitter;
std::shared_ptr<p2psp::SplitterIMS> splitter_ptr;
// True if splitter_ptr is SplitterIMS and no subclass
bool is_IMS_only;

void HandlerCtrlC(int s) {
  LOG("Keyboard interrupt detected ... Exiting!");

  // Say to daemon threads that the work has been finished,
  splitter_ptr->SetAlive(false);
}

void HandlerEndOfExecution() {
  // Wake up the "moderate_the_team" daemon, which is waiting in a recvfrom().
  splitter_ptr->SayGoodbye();

  // Wake up the "handle_arrivals" daemon, which is waiting in an accept().
  boost::asio::io_service io_service_;
  boost::system::error_code ec;
  boost::asio::ip::tcp::socket socket(io_service_);
  boost::asio::ip::tcp::endpoint endpoint(
      boost::asio::ip::address::from_string("127.0.0.1"),
      splitter_ptr->GetPort());

  socket.connect(endpoint, ec);

  // TODO: If "broken pipe" errors have to be avoided, replace the close method
  // with multiple receive calls in order to read the configuration sent by the
  // splitter
  socket.close();
}

bool HasParameter(const boost::program_options::variables_map& vm,
    const std::string& param_name, char min_magic_flags) {
  if (!vm.count(param_name)) {
    return false;
  }
  if (is_IMS_only || std::static_pointer_cast<p2psp::SplitterDBS>(
      splitter_ptr)->GetMagicFlags() < min_magic_flags) {
    ERROR("The parameter --" << param_name
        << " is not available for this splitter mode.");
    return false;
  }
  return true;
}

int main(int argc, const char *argv[]) {
  // Argument Parser
  boost::program_options::options_description desc(
      "This is the splitter node of a P2PSP team.  The splitter is in charge "
      "of defining the Set or Rules (SoR) that will control the team. By "
      "default, DBS (unicast transmissions) will be used");

  // TODO: strpe option should expect a list of arguments, not bool
  desc.add_options()("help,h", "Produce help message")("splitter_addr",
                     boost::program_options::value<std::string>(),
                     "IP address to serve (TCP) the peers. (Default = '{}')")(
      "buffer_size", boost::program_options::value<int>(),
      "size of the video buffer in blocks. Default = {}.")(
      "channel", boost::program_options::value<std::string>(),
      "Name of the channel served by the streaming source. Default = '{}'.")(
      "chunk_size", boost::program_options::value<int>(),
      "Chunk size in bytes. Default = '{}'.")(
      "header_size", boost::program_options::value<int>(),
      "Size of the header of the stream in chunks. Default = '{}'.")(
      "max_chunk_loss", boost::program_options::value<int>(),
      "Maximum number of lost chunks for an unsupportive peer. Makes sense "
      "only in unicast mode. Default = '{}'.")(
      "max_number_of_monitor_peers", boost::program_options::value<int>(),
      "Maximum number of monitors in the team. The first connecting peers will "
      "automatically become monitors. Default = '{}'.")(
      "mcast_addr", boost::program_options::value<std::string>(),
      "IP multicast address used to serve the chunks. Makes sense only in "
      "multicast mode. Default = '{}'.")(
      "port", boost::program_options::value<int>(),
      "Port to serve the peers. Default = '{}'.")(
      "source_addr", boost::program_options::value<std::string>(),
      "IP address or hostname of the streaming server. Default = '{}'.")(
      "source_port", boost::program_options::value<int>(),
      "Port where the streaming server is listening. Default = '{}'.")(
      "IMS", "Uses the IP multicast infrastructure, if available. IMS mode is "
      "incompatible with ACS, LRS, DIS and NTS modes.")(
      "NTS", "Enables NAT traversal.")(
      "ACS", "Enables Adaptative Chunk-rate.")(
      "LRS", "Enables Lost chunk Recovery")(
      "DIS", "Enables Data Integrity check.")(
      "strpe", "Selects STrPe model for DIS.")("strpeds", "Selects STrPe-DS model for DIS.")(
      "strpeds_majority_decision",
      "Sets majority decision ratio for STrPe-DS model.")(
      "strpe_log", boost::program_options::value<std::string>(),
      "Loggin STrPe & STrPe-DS specific data to file.")(
      "TTL", boost::program_options::value<int>(),
      "Time To Live of the multicast messages. Default = '{}'.");

  boost::program_options::variables_map vm;
  try {
    boost::program_options::store(boost::program_options::parse_command_line(argc, argv, desc), vm);
  } catch (std::exception &e) {

    // If the argument passed is unknown, print the list of available arguments
    std::cout << desc << "\n";
    return 1;
  }

  boost::program_options::notify(vm);

  if (vm.count("help")) {
    std::cout << desc << "\n";
    return 1;
  }

  is_IMS_only = false;
  if (vm.count("strpe")) {
    splitter_ptr.reset(new p2psp::SplitterSTRPE());
  } else if (vm.count("LRS")) {
    splitter_ptr.reset(new p2psp::SplitterLRS());
  } else if (vm.count("ACS")) {
    splitter_ptr.reset(new p2psp::SplitterACS());
  } else if (vm.count("NTS")) {
    splitter_ptr.reset(new p2psp::SplitterNTS());
  } else if (vm.count("IMS")) {
    is_IMS_only = true;
    splitter_ptr.reset(new p2psp::SplitterIMS());
  } else {
    splitter_ptr.reset(new p2psp::SplitterDBS());
  }

  if (vm.count("buffer_size")) {
    splitter_ptr->SetBufferSize(vm["buffer_size"].as<int>());
  }

  if (vm.count("channel")) {
    splitter_ptr->SetChannel(vm["channel"].as<std::string>());
  }

  if (vm.count("chunk_size")) {
    splitter_ptr->SetBufferSize(vm["chunk_size"].as<int>());
  }

  if (vm.count("header_size")) {
    splitter_ptr->SetHeaderSize(vm["header_size"].as<int>());
  }

  if (vm.count("port")) {
    splitter_ptr->SetPort(vm["port"].as<int>());
  }

  if (vm.count("source_addr")) {
    splitter_ptr->SetSourceAddr(vm["source_addr"].as<std::string>());
  }

  if (vm.count("source_port")) {
    splitter_ptr->SetSourcePort(vm["source_port"].as<int>());
  }

  // Parameters if splitter is not IMS
  if (HasParameter(vm, "max_chunk_loss", p2psp::Common::kDBS)) {
    std::shared_ptr<p2psp::SplitterDBS> splitter_dbs =
        std::static_pointer_cast<p2psp::SplitterDBS>(splitter_ptr);
    splitter_dbs->SetMaxChunkLoss(vm["max_chunk_loss"].as<int>());
  }

  if (HasParameter(vm, "max_number_of_monitor_peers", p2psp::Common::kDBS)) {
    std::shared_ptr<p2psp::SplitterDBS> splitter_dbs =
        std::static_pointer_cast<p2psp::SplitterDBS>(splitter_ptr);
    splitter_dbs->SetMonitorNumber(vm["max_number_of_monitor_peers"].as<int>());
  }

  // Parameters if STRPE
  if (HasParameter(vm, "strpe_log", p2psp::Common::kSTRPE)) {
    std::shared_ptr<p2psp::SplitterSTRPE> splitter_strpe =
        std::static_pointer_cast<p2psp::SplitterSTRPE>(splitter_ptr);
    splitter_strpe->SetLogging(true);
    splitter_strpe->SetLogFile(vm["strpe_log"].as<std::string>());
  }

  splitter_ptr->Start();

  LOG("         | Received  | Sent      | Number       losses/ losses");
  LOG("    Time | (kbps)    | (kbps)    | peers (peer) sents   threshold "
      "period kbps");
  LOG("---------+-----------+-----------+-----------------------------------.."
      ".");

  int last_sendto_counter = splitter_ptr->GetSendToCounter();
  int last_recvfrom_counter = splitter_ptr->GetRecvFromCounter();

  int chunks_sendto = 0;
  int kbps_sendto = 0;
  int kbps_recvfrom = 0;
  int chunks_recvfrom = 0;
  std::vector<boost::asio::ip::udp::endpoint> peer_list;

  // Listen to Ctrl-C interruption
  struct sigaction sigIntHandler;
  sigIntHandler.sa_handler = HandlerCtrlC;
  sigemptyset(&sigIntHandler.sa_mask);
  sigIntHandler.sa_flags = 0;
  sigaction(SIGINT, &sigIntHandler, NULL);

  std::shared_ptr<p2psp::SplitterDBS> splitter_dbs;
  if (!is_IMS_only) { // GetPeerList is only in DBS and derivated classes
    splitter_dbs = std::static_pointer_cast<p2psp::SplitterDBS>(splitter_ptr);
  }
  std::shared_ptr<p2psp::SplitterACS> splitter_acs;
  if (!is_IMS_only && splitter_dbs->GetMagicFlags() >= p2psp::Common::kACS) {
    splitter_acs = std::static_pointer_cast<p2psp::SplitterACS>(splitter_ptr);
  }
  while (splitter_ptr->isAlive()) {
    boost::this_thread::sleep(boost::posix_time::milliseconds(1000));
    chunks_sendto = splitter_ptr->GetSendToCounter() - last_sendto_counter;
    kbps_sendto = (chunks_sendto * splitter_ptr->GetChunkSize() * 8) / 1000;
    chunks_recvfrom =
        splitter_ptr->GetRecvFromCounter() - last_recvfrom_counter;
    kbps_recvfrom = (chunks_recvfrom * splitter_ptr->GetChunkSize() * 8) / 1000;
    last_sendto_counter = splitter_ptr->GetSendToCounter();
    last_recvfrom_counter = splitter_ptr->GetRecvFromCounter();

    LOG("|" << kbps_recvfrom << "|" << kbps_sendto << "|");
    // LOG(_SET_COLOR(_CYAN));

    if (!is_IMS_only) { // GetPeerList is only in DBS and derivated classes
      peer_list = splitter_dbs->GetPeerList();
      LOG("Size peer list: " << peer_list.size());

      std::vector<boost::asio::ip::udp::endpoint>::iterator it;
      for (it = peer_list.begin(); it != peer_list.end(); ++it) {
        // _SET_COLOR(_BLUE);
        LOG("Peer: " << *it);
        // _SET_COLOR(_RED);

        LOG(splitter_dbs->GetLoss(*it) << "/" << chunks_sendto << " "
                                       << splitter_dbs->GetMaxChunkLoss());

        if (splitter_dbs->GetMagicFlags() >= p2psp::Common::kACS) { // If is ACS
          // _SET_COLOR(_YELLOW);
          LOG(splitter_acs->GetPeriod(*it));
          // _SET_COLOR(_PURPLE)
          LOG((splitter_acs->GetNumberOfSentChunksPerPeer(*it) *
               splitter_acs->GetChunkSize() * 8) /
              1000);
          splitter_acs->SetNumberOfSentChunksPerPeer(*it, 0);
        }
      }
    }
  }

  LOG("Ending");
  HandlerEndOfExecution();

  return 0;
}
